File: DetectObjects.m
function [o] = DetectObjects(ImageFile,Confidence)
%DETECTOBJECTS Detect objects using system call
%   Objects = table of object classes and bounding boxes
%       category : car
%       startX : 430
%       startY : 316
%       endX : 706
%       endY :494
%       confidence : 0.956794
%   ImageFile : full path to the image file
%   Confidence : level of confidence for the objects included, 0...1
    % Spawn the Python application
    command = "python detect_objects.py --image ";
    command = strcat(command,ImageFile);
    command = strcat(command," --confidence ");
    command = strcat(command,num2str(Confidence));
    [status,cmdout] = system(command);
    if status ~= 0; o=null; end
    % Parse the results
    c=textscan(cmdout,'%s %d %d %d %d %f');
    % Change the results to a struct array
    [h,~]=size(c{1});
    for i=1:h
        o(i).class=c{1}(i);
        o(i).startX=c{2}(i);
        o(i).startY=c{3}(i);
        o(i).endX=c{4}(i);
        o(i).endY=c{5}(i);
        o(i).confidence=c{6}(i);
    end
end

File: detect_objects.py
# USAGE
# python detect_objects.py --image images/example_01.jpg --confidence 0.2

# import the necessary packages
import numpy as np
import argparse
import cv2

# construct the argument parse and parse the arguments
ap = argparse.ArgumentParser()
ap.add_argument("-i", "--image", required=True,
	help="path to input image")
ap.add_argument("-c", "--confidence", type=float, default=0.2,
	help="minimum probability to filter weak detections")
args = vars(ap.parse_args())

# initialize the list of class labels MobileNet SSD was trained to
# detect, then generate a set of bounding box colors for each class
CLASSES = ["background", "aeroplane", "bicycle", "bird", "boat",
	"bottle", "bus", "car", "cat", "chair", "cow", "diningtable",
	"dog", "horse", "motorbike", "person", "pottedplant", "sheep",
	"sofa", "train", "tvmonitor"]

# load our serialized model from disk
net = cv2.dnn.readNetFromCaffe("MobileNetSSD_deploy.prototxt.txt", "MobileNetSSD_deploy.caffemodel")

# load the input image and construct an input blob for the image
# by resizing to a fixed 300x300 pixels and then normalizing it
# (note: normalization is done via the authors of the MobileNet SSD
# implementation)
image = cv2.imread(args["image"])
(h, w) = image.shape[:2]
blob = cv2.dnn.blobFromImage(cv2.resize(image, (300, 300)), 0.007843, (300, 300), 127.5)

# pass the blob through the network and obtain the detections and
# predictions
net.setInput(blob)
detections = net.forward()

# loop over the detections
for i in np.arange(0, detections.shape[2]):
	# extract the confidence (i.e., probability) associated with the
	# prediction
	confidence = detections[0, 0, i, 2]

	# filter out weak detections by ensuring the `confidence` is
	# greater than the minimum confidence
	if confidence > args["confidence"]:
		# extract the index of the class label from the `detections`,
		# then compute the (x, y)-coordinates of the bounding box for
		# the object
		idx = int(detections[0, 0, i, 1])
		box = detections[0, 0, i, 3:7] * np.array([w, h, w, h])
		(startX, startY, endX, endY) = box.astype("int")
		# display the prediction
		print("%s %d %d %d %d %f" % (CLASSES[idx],startX,startY,endX,endY,confidence))